이슈 리스트 페이지에서 서버에 api 요청 및 useReducer로 상태 관리

yejineee opened the pull request:
- 서버의 주소는 매 번 반복되므로, 오류를 줄이고, 편리함을 높이기 위해
axios의 instance를 생성하여 기본 URL을 설정하였다.

- axios의 response에서 매 번 data를 받아와야 하므로, interceptor하여
data를 바로 넘겨줄 수 있도록 설정하였다.

- IssueListPage에서 api요청으로 issue들을 받아서, 자식 컴포넌트에게
전달하여 렌더링하도록 해야 한다.
따라서 api를 요청한 후, issue들을 받고, 상태 변경을 useReducer를 사용하도록 하였다.

- 이슈 리스트 페이지에서 api요청을 한 후 상태 변경을 useReducer로 하기 때문에, 기존에 useState로 관리하던 것을 useReducer로 바꿔줘야 했다. 따라서 IssueListContainer안에 있는 컴포넌트에서 클릭 이벤트 발생 시
상태 변경을 reducer가 하도록 수정하였다.

yejineee committed:
Feat: 이슈 리스트 컴포넌트 생성 (#30)

* Refactor: App.js에서 가장 첫 페이지를 이슈 리스트 페이지로 설정

로그인 후, 가장 먼저 보이는 페이지를 이슈 리스트 페이지이므로
해당 페이지 컴포넌트를 가져와서 렌더링하는 것으로 수정하였다.

* Style: 이슈 리스트에서 쓰이는 컬러 추가

이슈 리스트에서 쓰이는 파란색 3개를 추가하였다.

* Feat: 리스트 헤더 컴포넌트 추가

리스트 헤더는 라벨 페이지, 이슈 리스트 페이지, 마일스톤 리스트 페이지 등
여러 페이지의  헤더에 공통적으로 쓰인다. 따라서 스타일을 지정한 리스트 컴포넌트를
추가하였다.

* Feat: 이슈 리스트 컨테이너 컴포넌트 추가

첫 페이지의 이슈 리스트 헤더와 목록들을 보여주는
컴포넌트를 추가하였다.
페이지부터, setState함수를 가져와서 이슈  헤더 클릭 이벤트와
이슈 바디 클릭 이벤트 핸들러를 따로 정의하였다.
이렇게 함으로써, 어떤 버튼이 클릭되었는지에 대한 분기 처리를 하지
않아도 되도록 하였다.

* Feat: 이슈 리스트 페이지 컴포넌트 추가

이슈 리스트 페이지 컴포넌트는 이슈 필터 컴포넌트와, 이슈 리스트 컴포넌트를
자식으로 갖고 있다.
두 컴포넌트에서 공통으로 state를 사용하기 때문에, 이 페이지 컴포넌트에서
state를 초기화하고, state와 setState 함수를 자식에게 넘겨주도록 한다.

Co-authored-by: yejineee <21700431@handong.edu>

rolled-potatoes committed:
Feat : issue 더미 api 생성 (#31)

임시로 보여줄 데이터를 json으로 읽어와서 반환 함.

rolled-potatoes committed:
Docs : issue 더미 api 명세서 작성

Commit:
Chore: axios baseURL 설정 및 response interceptor 설정

- 서버의 주소는 매 번 반복되므로, 오류를 줄이고, 편리함을 높이기 위해
axios의 instance를 생성하여 기본 URL을 설정하였다.

- axios의 response에서 매 번 data를 받아와야 하므로, interceptor하여
data를 바로 넘겨줄 수 있도록 설정하였다.

Commit:
Chore: 모든 이슈를 서버로부터 받아오는 axios 설정

서버로부터 api요청하는 코드를 apis 폴더에 저장한다.
그 중 api/issue.js는 이슈 모델과 관련하여 api요청하는 코드를 모아둘 것이다.
이슈의 모든 모델들을 가져오는 axios를 추가하였다.

Commit:
Chore: issue의 useReducer에서 사용되는 action type 상수 저장

useReducer에서 사용되는 action의 type을 상수로 지정하여,
코드 오류를 줄이고, 생산성을 높이도록 하였다.

Commit:
Feat: issue의 useReducer에서 쓰이는 reducer 정의

issue 페이지에서 쓰일 reducer를 정의하였다.
- FETCH_SUCCESS:
    - fetch 성공시 issues state를 새롭게 정의한다.
- FETCH_ERROR:
    - Fetch에 실패시, error state를 정의한다.
- CHECK_ALL_ISSUE_HANDLER:
    - 이슈 헤더에 있는 체크박스 클릭 시, 전체 issue 상태를 바꿔준다.
- CHECK_ISSUE_HANDLER:
    - 각 이슈 옆에 있는 체크박스 클릭 시, 해당 issue의 상태를 바꿔주고, 필요시 전체 issue의 상태를 바꿔준다.

Commit:
Refactor: 이슈 리스트 컴포넌트의 상태관리 방법을 useReducer로 바꿈

useState로 상태를 관리할 경우, 여러 컴포넌트에서 상태를 변경하게 되어
한 눈에 알아보기 힘든 점이 있었다.
또한, 이슈 리스트 페이지에서 api요청을 한 후 상태 변경을 useReducer로 하기 때문에, 기존에 useState로 관리하던 것을 useReducer로 바꿔줘야 했다.
따라서 IssueListContainer안에 있는 컴포넌트에서 클릭 이벤트 발생 시
상태 변경을 reducer가 하도록 수정하였다.

Commit:
Feat: IssueListPage에서 api요청하여 데이터 받는 것 구현

IssueListPage에서 api요청으로 issue들을 받아서, 자식 컴포넌트에게
전달하여 렌더링하도록 해야 한다.
따라서 api를 요청한 후, issue들을 받고, 상태 변경을 useReducer를 사용하도록 하였다.

Commit:
Chore: axios 설치

서버에 api요청 시 사용하는 라이브러리로 axios를 선택하여, 설치하였다.

Commit:
Chore: 프론트 eslint 규칙 추가

changheedev commented:
늦은시간까지... 고생많으셨네요 👍 

issue-list 브랜치로 pr이 올라왔는데 혹시 master로 올라갈 pr일까요??

changheedev commented on front/src/routes/IssueList/IssueListPage.js:
```diff
@@ -1,20 +1,49 @@
   });
 
+  const fetchIssue = async () => {
+    try {
+      const response = await issueApi.getIssues();
+      dispatch({ type: FETCH_SUCCESS, issueItems: response });
+    } catch (e) {
```
issue를 받아온 결과이니까 response 보다는 issues 같은 변수명이 더 명확하고 어울릴 것 같습니다.

yejineee replied:
넹...명확한 이름으로 수정하겠슴니다!

rolled-potatoes commented on front/src/routes/IssueList/issueHook.js:
```diff
@@ -0,0 +1,70 @@
+}
+
+const setAllIssueChecked = (issues, value) =>
+  issues.map((issue) => {
+    return { ...issue, checked: value };
+  });
```
중괄호를 지우면 리턴도 지울 수 있을 것 같습니다.

yejineee replied:
넵...ㅎㅎㅎ 간결하게 바꿔볼게요!

Commit:
Style: map에서 중괄호 삭제, 리턴 받는 변수명 변경

좀 더 명확한 네이밍을 위해, 서버로 부터 받아온 데이터를 issues라고
명명하였다.
map에서 하나의 statement만 있을 때, 중괄호를 삭제할 수 있으므로 삭제하였다.
